packages/serverpod/lib/src/database/migrations/migration_manager.dart

```diff
@@ -141,42 +141,85 @@ class MigrationManager {
       return null;
     }
 
-    var appliedRepairMigration = await DatabaseMigrationVersion.db.findFirstRow(
-        session,
-        where: (t) =>
-            t.module.equals(MigrationConstants.repairMigrationModuleName));
+    String? appliedVersionName = repairMigration.versionName;
+    await _withMigrationLock(session, () async {
+      var appliedRepairMigration = await DatabaseMigrationVersion.db
+          .findFirstRow(session,
+              where: (t) => t.module
+                  .equals(MigrationConstants.repairMigrationModuleName));
+
+      if (appliedRepairMigration != null &&
+          appliedRepairMigration.version == repairMigration.versionName) {
+        appliedVersionName = null;
+        return;
+      }
 
-    if (appliedRepairMigration != null &&
-        appliedRepairMigration.version == repairMigration.versionName) {
-      return null;
-    }
+      await session.db.unsafeSimpleExecute(
+        repairMigration.sqlMigration,
+      );
 
-    await session.db.unsafeSimpleExecute(
-      repairMigration.sqlMigration,
-    );
-    return repairMigration.versionName;
+      await _updateState(session);
+    });
+
+    return appliedVersionName;
   }
 
   /// Migrates all modules to the latest version.
   ///
   /// Returns the migrations applied.
   /// Returns null if latest version was already installed.
   Future<List<String>?> migrateToLatest(Session session) async {
-    var latestVersion = getLatestVersion();
+    List<String>? migrationsApplied = [];
 
-    var moduleName = session.serverpod.serializationManager.getModuleName();
+    await _withMigrationLock(session, () async {
+      await _updateState(session);
+      var latestVersion = getLatestVersion();
 
-    if (isVersionInstalled(moduleName, latestVersion)) {
-      return null;
-    }
+      var moduleName = session.serverpod.serializationManager.getModuleName();
 
-    var installedVersion = getInstalledVersion(moduleName);
+      if (isVersionInstalled(moduleName, latestVersion)) {
+        migrationsApplied = null;
+        return;
+      }
 
-    return await _migrateToLatestModule(
-      session,
-      latestVersion: latestVersion,
-      fromVersion: installedVersion,
-    );
+      var installedVersion = getInstalledVersion(moduleName);
+
+      migrationsApplied = await _migrateToLatestModule(
+        session,
+        latestVersion: latestVersion,
+        fromVersion: installedVersion,
+      );
+      await _updateState(session);
+    });
+
+    return migrationsApplied;
+  }
+
+  Future<void> _withMigrationLock(
+    Session session,
+    Future<void> Function() action,
+  ) async {
+    const String lockName = 'serverpod_migration_lock';
+
+    /// Use a transaction to ensure that the advisory lock is retained
+    /// until the transaction is completed.
```

Comment:
```suggestion
    /// until the transaction is completed.
    ///
    /// The transaction ensures that the session used for acquiring the
    /// lock is kept alive in the underlying connection pool, and that we
    /// can later use that exact same session for releasing the lock.
    /// The transaction is thus only used to get the desired behavior from
    /// the database driver, and does not have any effect on the Postgres level.
```

I think since this is just an implementation comment, we can be very verbose / comprehensive, so that this does not get message up in the future by someone (like me 10 minutes ago) trying to optimize this ;)

Reply:
Thanks for the suggestion! I'll use this straight of :)